Enforced maximum amount of students for CCX

giocalitri opened the pull request:
CCX overrides course `max_student_enrollments_allowed` and set it to 200 during CCX creation.

Currently Custom Courses for Edx do not have a way to limit the students that can be enrolled and only inherit the limit (if any) from the parent course.
This PR introduces this limit and it defaults it to 200 for any new CCX. This PR does not address the modification of this value.

The way `max_student_enrollments_allowed` is overridden is the same way other CCX fields, such as `start` and `due` dates, override the parent course field.

The `200` limit has been made configurable.

Please see https://github.com/mitocw/edx-platform/issues/122 for more info.

edx-webhook commented:
Thanks for the pull request, @giocalitri! I've created [OSPR-929](https://openedx.atlassian.net/browse/OSPR-929) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).
We can't start reviewing your pull request until you've   added yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS) file . Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

amir-qayyum-khan commented on lms/envs/common.py:
```diff
@@ -2701,3 +2701,10 @@
+# This is an arbitrary hard limit.
+# The reason we introcuced this number is because we do not want the CCX
+# to compete with the MOOC.
+CCX_MAX_STUDENTS_ALLOWED = 200
```
FYI, I am not sure what kind of string/constants edX support in site configuration files. I got below feedback when i put maximum number of students per page in common.py. They suggest me to put constant in app.
 
> this is also probably not the right place for this setting to live. We may also want to just pick a number and not allow it to be site configurable, because this configuration file is 2682 (!!!!) lines long, which is insane.

sarina replied:
Configuring the number of items to show on a page is a UX decision and it's a bit insane to make all those things site wide configurable.

Limiting the size of a CCX potentially makes more sense as a sitewide configuration option. I'm curious what product or marketing decisions have led to a limit of 200. If it's truly arbitrary, I think it merits a bit more discussion. We shouldn't be adding features without having good justifications for the decisions we're making.

sarina replied:
@pdpinch ^ I'm still curious/concerned about this. Have you had time to figure it out?

amir-qayyum-khan commented on lms/djangoapps/ccx/views.py:
```diff
@@ -433,9 +440,17 @@ def ccx_invite(request, course, ccx=None):
+        email_params = get_email_params(course, auto_enroll, course_key=course_key, display_name=ccx.display_name)
+        if action == 'Enroll':
+            # check if there are spots left
+            if CourseEnrollment.objects.is_course_full(ccx_course_overview):
```
@giocalitri  i think you need to update logic of ```ccx_invite```, 
If ```is_course_full is True``` then there is no point of validate student , just check in start if  mode is ```enroll``` and  ```is_course_full is True``` then inform coach that he cannot add student.

amir-qayyum-khan replied:
@giocalitri  i think you need to update logic of ```ccx_invite```, 
If ```is_course_full is True``` then there is no point of validate student , just check in start of loop if  mode is ```enroll``` and  ```is_course_full is True``` then inform coach that he cannot add student.

giocalitri replied:
You are right: will do

amir-qayyum-khan commented on lms/djangoapps/ccx/views.py:
```diff
@@ -507,12 +521,19 @@ def ccx_student_management(request, course, ccx=None):
-            enroll_email(course_key, email, auto_enroll=True, email_students=False)
+            # check if there are spots left
+            ccx_course_overview = CourseOverview.get_from_id(course_key)
+            if CourseEnrollment.objects.is_course_full(ccx_course_overview):
```
@giocalitri  i think you need to update logic of ```ccx_student_management```, 
If ```is_course_full is True``` then there is no point of validate student , just check in start of function if  mode is ```enroll``` and  ```is_course_full is True``` then inform coach that he cannot add student.

giocalitri replied:
Will do

amir-qayyum-khan commented on lms/envs/aws.py:
```diff
@@ -669,6 +669,7 @@
     FIELD_OVERRIDE_PROVIDERS += (
         'ccx.overrides.CustomCoursesForEdxOverrideProvider',
     )
+CCX_MAX_STUDENTS_ALLOWED = ENV_TOKENS.get('CCX_MAX_STUDENTS_ALLOWED', CCX_MAX_STUDENTS_ALLOWED)
```
@giocalitri  Need to be sure if you guys want this constant as side configureable? see my below comment which i receive on one of my PR.

giocalitri replied:
I see some value about having this value configurable, but I have not strong arguments about it. @pdpinch, what do you think?

pdpinch replied:
I think whether or not to make this configurable needs to be edX's decision. I'm fine as long as it isn't more than 200. 

@sarina let us know what you think or whose opinions we need to bring in.

sarina replied:
If we're going to improve the gradebook to have server side pagination,
ideally the performance will be good enough to kick this up to at least 500
or even 1000 - which frankly is reasonable for an on campus course. more
than 1000 seems ridiculous, though.

On Wed, Nov 4, 2015 at 12:07 PM, Peter Pinch <notifications@github.com>
wrote:

> In lms/envs/aws.py
> <https://github.com/edx/edx-platform/pull/10483#discussion_r43908184>:
>
> > @@ -669,6 +669,7 @@
> >      FIELD_OVERRIDE_PROVIDERS += (
> >          'ccx.overrides.CustomCoursesForEdxOverrideProvider',
> >      )
> > +CCX_MAX_STUDENTS_ALLOWED = ENV_TOKENS.get('CCX_MAX_STUDENTS_ALLOWED', CCX_MAX_STUDENTS_ALLOWED)
>
> I think whether or not to make this configurable needs to be edX's
> decision. I'm fine as long as it isn't more than 200.
>
> @sarina <https://github.com/sarina> let us know what you think or whose
> opinions we need to bring in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/10483/files#r43908184>.
>

sarina replied:
Sorry - obviously I replied here via email and was context switching from a different PR. Why do you need this to be 200 or less?

sarina replied:
n/m I see the comments in another file.

pdpinch replied:
I'm going to discuss this with edX product to confirm the default maximum.

amir-qayyum-khan commented:
@giocalitri so work this is only for new ccx course?

giocalitri commented:
@amir-qayyum-khan
> @giocalitri so work this is only for new ccx course?

Yes.

pdpinch commented:
@sarina I'm realizing that we will need to make some documentation updates for this. What's the best way to do that -- do we just open PRs on the docs repo, or do we need to open JIRA tickets first?

sarina commented:
@pdpinch You can open PRs directly onto the docs repo - jira tickets will get made for them automagically.

giocalitri commented:
jenkins run js

giocalitri commented:
@amir-qayyum-khan I addressed your comments: please review again

amir-qayyum-khan commented on lms/djangoapps/ccx/tests/test_views.py:
```diff
@@ -427,6 +434,40 @@ def test_enroll_member_student(self):
+        ccx_course_key = CCXLocator.from_course_locator(self.course.id, ccx.id)
+        # create some users
+        students = [
+            UserFactory.create(username="test{0}".format(seq), is_staff=False) for seq in range(3)
```
Do you need this code some where ```username="test{0}".format(seq)```. UserFactory manages its sequence of creation of multiple users. Please see below code:
https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/tests/factories.py#L76
https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/tests/factories.py#L77

giocalitri replied:
did not know: will fix the code

amir-qayyum-khan commented on lms/djangoapps/ccx/tests/test_views.py:
```diff
@@ -606,6 +647,44 @@ def test_manage_remove_single_student(self):
+        ccx = self.make_ccx(max_students_allowed=1)
+        ccx_course_key = CCXLocator.from_course_locator(self.course.id, ccx.id)
+        students = [
+            UserFactory.create(username="test{0}".format(seq), is_staff=False) for seq in range(2)
```
same question as above?

giocalitri replied:
same answer as above

amir-qayyum-khan commented on lms/djangoapps/ccx/tests/test_views.py:
```diff
@@ -606,6 +647,44 @@ def test_manage_remove_single_student(self):
+        )
+        # try to enroll the second student without success
+        # enroll the first student
+        data = {
```
please refactor this test use ddt.

for example see https://github.com/edx/edx-platform/blob/master/lms/djangoapps/ccx/tests/test_views.py#L820

giocalitri replied:
will look into this: not sure what `ddt` does yet

amir-qayyum-khan replied:
@giocalitri  FYI http://ddt.readthedocs.org/en/latest/

giocalitri replied:
I do not think I need to use `ddt`: what I'm trying to do is to demonstrate that if I have the limit set to 1 and I enroll 2 students, the first will work and the second will not. 
Why `ddt` should help me? I'm not trying to run the same test with two different values.

sarina replied:
Yeah I don't think you neeed ddt here. but you should update the comment at L710

sarina replied:
However, this test and `test_enroll_single_non_user_student` are really similar. You can probably at minimum refactor these tests to use helper functions. And keep DDT in mind - we really want to prevent writing the same code in multiple places.

amir-qayyum-khan commented on lms/djangoapps/ccx/views.py:
```diff
@@ -422,20 +446,17 @@ def ccx_invite(request, course, ccx=None):
+    email_params = get_email_params(course, auto_enroll, course_key=course_key, display_name=ccx.display_name)
+
+    if action == 'Enroll':
+        for identifier in identifiers:
```
why not do this to save code duplication, i mean using for loop for both enroll and unenroll?
```
for identifier in identifiers:
  if action == 'Enroll':
     if ccx has not filled and user is valid
        add user
  else if action == "Unenroll":
     unenroll user
```
or you can create new methods to enroll and unenroll single or multiple users.
```
def enroll_students() # Do enroll validations here 
def unenroll_student() 
if action == 'Enroll':
     enroll_students()
  else if action == "Unenroll":
     unenroll_student()
```

giocalitri replied:
>@giocalitri why not do this to save code duplication, i mean using for loop for both enroll and unenroll?
```
for identifier in identifiers:
  if action == 'Enroll':
     if ccx has not filled and user is valid
        add user
  else if action == "Unenroll":
     unenroll user
```

I think this approach is wrong: there can be only one `action` per submission and it is useless to check it at each loop.

giocalitri replied:
> Only suggestion: you can also create new methods to enroll and unenroll single or multiple users.
FYI in my PR i am adding email functionality in ccx_student_management mitocw#135
```
def enroll_students() # Do enroll validations here 
def unenroll_student() 
if action == 'Enroll':
     enroll_students()
  else if action == "Unenroll":
     unenroll_student()
```

I'm not sure I understood what you mean.

amir-qayyum-khan replied:
i am saying you can create 2 methods and move enrolment and unenrolment code their.
because we are doing this in two methods ```ccx_invite and ccx_student_management```.
You can create new methods mention above which will handle enrolment and unenrolment for single and multiple users.

sarina commented on lms/djangoapps/ccx/views.py:
```diff
@@ -59,6 +61,18 @@
+    """
+    Custom Exception for full ccx courses
+    """
+
```
Generally we explicitly add in a `pass`. see openedx/core/djangoapps/content/course_structures/api/v0/errors.py

giocalitri replied:
@sarina will do immediately

giocalitri commented:
@amir-qayyum-khan I think I addressed all your comment but the one about `ddt` (see my comment there)

@pdpinch I refactored the code for enrollment and now the right side of the CCX Coach automatically enrolls valid emails too. The only difference left with the left side is that no email is sent.
Please test.

sarina commented on lms/djangoapps/ccx/views.py:
```diff
@@ -407,6 +424,47 @@ def ccx_schedule(request, course, ccx=None):  # pylint: disable=unused-argument
+def _get_valid_student_email(identifier):
+    """
+    helper function to get an user email
+    from an identifier and validate it
```
This docstring isn't super helpful, it doesn't explain what an identifer is, and what values are returned or exceptions raised. See https://github.com/edx/edx-platform/wiki/Python-Guidelines#follow-pep-257

giocalitri replied:
I did not add a proper description in this function because I declared it "private", but I will take care of that

sarina commented on lms/djangoapps/ccx/views.py:
```diff
@@ -489,30 +511,28 @@ def ccx_student_management(request, course, ccx=None):
+    email_params = get_email_params(course, auto_enroll=True, course_key=course_key, display_name=ccx.display_name)
+
+    if action == 'add':
+        # by decree, no emails sent to students added this way
```
? Why?

giocalitri replied:
I guess there is an historical reason: I just refactored the code to use the same helper functions, but I did not change this part of the logic.

sarina replied:
Yeah it would just be better to have better reasoning in here. This is a living codebase, the developers working on this code may not be here tomorrow. It's better to have comments that explain and contextualize the code, that anyone new to the codebase can read and understand. This is cryptic, and not helpful. It would almost be better to just say,

`Students added (this way) don't get emailed, and are auto-enrolled`

where instead of saying "(this way)" you actually say what "this way" is.

pdpinch replied:
Let's standardize this and do the same thing in both cases. @amir-qayyum-khan  already started a PR for adding the UI: https://github.com/mitocw/edx-platform/pull/135

@giocalitri can you figure out how best to reconcile the two PRs?

giocalitri replied:
@pdpinch: do you want I merge his work in this PR or just make this PR compatible?

sarina commented on lms/djangoapps/ccx/views.py:
```diff
@@ -489,30 +511,28 @@ def ccx_student_management(request, course, ccx=None):
+            _enroll_students(
+                identifiers=[student_id], course_key=course_key, email_students=False, email_params=email_params)
+        except (CCXCourseFullException, CCXUserValidationException) as exp:
+            error_message = "{0}".format(exp)
```
can you instead make it `log.info("%s", exp)`? See http://stackoverflow.com/a/4149231/2669626

giocalitri replied:
I return the same error message to the UI, so I decided to format it only once

sarina commented on lms/djangoapps/ccx/views.py:
```diff
@@ -489,30 +511,28 @@ def ccx_student_management(request, course, ccx=None):
+            _unenroll_students(
+                identifiers=[student_id], course_key=course_key, email_students=False, email_params=email_params)
+        except CCXUserValidationException as exp:
+            error_message = "{0}".format(exp)
```
Same here re lazy logging

sarina commented on lms/djangoapps/ccx/views.py:
```diff
@@ -489,30 +511,28 @@ def ccx_student_management(request, course, ccx=None):
-            unenroll_email(course_key, email, email_students=False)
-    else:
+    error_message = None
+    email_params = get_email_params(course, auto_enroll=True, course_key=course_key, display_name=ccx.display_name)
```
why do you bother getting the email params if you don't use them?

giocalitri commented:
@sarina: I refactored the code for the views and their tests. Now the views in CCX have exactly the same behavior. The reason I did not merge the views is because I did not want to make the work in the UI part of this PR (with an explosion of scope).

sarina commented:
I'm :+1: on this change but I still have open questions around 200 per ccx maximum limit.

However I'm happy to let this PR go through and have that discussion out of band.

@pdpinch when this is tested and approved on your side, go ahead and merge.

amir-qayyum-khan commented:
:+1: , Need to rebase

giocalitri commented:
Rebased

sarina commented:
@giocalitri Can you add yourself to the AUTHORS file? Please do it in a separate PR. (also would you mind adding any other contributors from your team who aren't in there)

giocalitri commented:
@sarina I am already there (9th from the bottom): I already had few PRs merged.

sarina commented:
@giocalitri huh, but the bot doesn't recognize that you're in the list.

sarina commented:
I guess maybe you opened multiple PRs at the same time, and weren't in authors when this was opened. If it happens on your next PR, then I'll consider the bot broken.

giocalitri commented:
Thanks!
